Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -154,31 +159,117 @@ namespace ROS2
+            pfClassId.name = "class_id";
+            pfClassId.offset = message.point_step;
+            pfClassId.datatype = sensor_msgs::msg::PointField::UINT8;
+            pfClassId.count = 1;
+            message.fields.push_back(pfClassId);
+            message.point_step += sizeof(uint8_t);
+        }
+
+        const auto sizeInBytes = pointCount * message.point_step;
         message.data.resize(sizeInBytes);
+
+        float x, y, z;
+        for (int i = 0; i < pointCount; ++i)
+        {
+            // to avoid alignment issues, we copy the data field by field
+            x = lastScanResults.m_points[i].GetX();
+            y = lastScanResults.m_points[i].GetY();
+            z = lastScanResults.m_points[i].GetZ();
+
+            memcpy(&message.data[i * message.point_step], &x, sizeof(float));
+            memcpy(&message.data[i * message.point_step + sizeof(float)], &y, sizeof(float));
+            memcpy(&message.data[i * message.point_step + 2 * sizeof(float)], &z, sizeof(float));
+            uint32_t nextFieldOffset = 3 * sizeof(float);
+
+            if (lastScanResults.m_ids.has_value())
+            {
+                memcpy(&message.data[i * message.point_step + nextFieldOffset], &lastScanResults.m_ids.value()[i], sizeof(int32_t));
+                nextFieldOffset += sizeof(int32_t);
+            }
+
+            if (publishSegmentation)
+            {
+                const AZ::Color color = colorLookupTable[lastScanResults.m_classes.value()[i]];
+                for (int j = 0; j < 3; j++)
+                {
+                    float chanelValue = color.GetElement(j);
+                    memcpy(&message.data[i * message.point_step + nextFieldOffset], &chanelValue, sizeof(float));
+                    nextFieldOffset += sizeof(float);
+                }
+                memcpy(&message.data[i * message.point_step + nextFieldOffset], &lastScanResults.m_classes.value()[i], sizeof(uint8_t));
+                nextFieldOffset += sizeof(uint8_t);
+            }
+        }
+        message.row_step = message.width * message.point_step;
         AZ_Assert(message.row_step * message.height == sizeInBytes, "Inconsistency in the size of point cloud data");
-        memcpy(message.data.data(), lastScanResults.m_points.data(), sizeInBytes);
+
         m_pointCloudPublisher->publish(message);
+
+        if (publishSegmentation && m_segmentationClassesPublisher)
+        {
+            vision_msgs::msg::LabelInfo segmentationClasses;
+            for (const auto&segmentation_class : m_lidarCore.m_lidarConfiguration.m_segmentationClasses)
+            {
+                vision_msgs::msg::VisionClass visionClass;
+                visionClass.class_id = segmentation_class.m_classId;
+                visionClass.class_name = segmentation_class.m_className.c_str();
+                segmentationClasses.class_map.push_back(visionClass);
+            }
+            m_segmentationClassesPublisher->publish(segmentationClasses);
+        }
```

Comment:
The method is too large and error-prone (risky to change) due to multiple / hand-crafted arithmetics on sizes & point_steps, and too many ifs.

I would like to have it as a generic handling of extra fields (as more will come) rather than a specifically named ones.

Perhaps seperate out creation of all sensor_msgs::msg::PointField and then just process them here.

The best flow would be to let the user chose the lidar implementation, following that let them pick from supported fields (x, y, z, intensity, class_id, ...), and handle that as modular (separate out features) + generic (handle them in an opaque manner).

Reply:
splited into lambdas